enarx-keep-sev-shim/src/lib.rs

```diff
@@ -0,0 +1,121 @@
+// SPDX-License-Identifier: Apache-2.0
+
+//! Shared components for the shim and the loader
+//! # Loader
+//!
+//! The loader calls [`BootInfo::calculate`] to get the offset for the shim and the code.
+//!
+//! The loader starts the virtual machine and jumps to the shim entry point.
+//!
+//! The shim expects the following registers:
+//! * `%rdi` = `SYSCALL_PHYS_ADDR`, address of the page, where the loader placed a copy of `BootInfo`
+//!            and which is used later on for the communication with the shim.
+//! * `%rsi` = the start address of the shim memory (contents of `BootInfo.shim.start`)
+//! * `%rip` = the address of the shim entry point taken from the elf header
+//!
+//! Although `%rsi` is redundant, it makes the initial startup function of the `shim` much easier.
+//!
+//! # Shim
+//!
+//! The shim sets the unencrypted flag for the page at `SYSCALL_PHYS_ADDR` and uses that page
+//! for further communication with the host.
+//!
+//! The `setup` area must not be touched, unless the shim sets up the page tables,
+//! the GDT and the IDT. After that the prefix area is used as free memory except for the pages
+//! to communicate with the host.
+//!
+//! To proxy a syscall to the host, the shim triggers a `#VMEXIT` via I/O on the
+//! [`SYSCALL_TRIGGER_PORT`].
+//!
+//! [`BootInfo::calculate`]: struct.BootInfo.html#method.calculate
+//! [`SYSCALL_TRIGGER_PORT`]: constant.SYSCALL_TRIGGER_PORT.html
+
+#![no_std]
+#![deny(clippy::all)]
+#![deny(clippy::integer_arithmetic)]
+#![deny(missing_docs)]
+
+/// I/O port used to trigger a `#VMEXIT`
+///
+/// FIXME: might change to another mechanism in the future
+pub const SYSCALL_TRIGGER_PORT: u16 = 0xFF;
+
+use span::{Line, Span};
+
+// page align
+const ALIGN: usize = 0x10_0000;
+
+#[inline(always)]
+const fn lower(value: usize, boundary: usize) -> usize {
+    #![allow(clippy::integer_arithmetic)]
+    // integer_arithmetic checked
+    value / boundary * boundary
+}
+
+#[inline(always)]
+fn raise(value: usize, boundary: usize) -> Option<usize> {
+    value
+        .checked_add(boundary)
+        .map(|v| v.wrapping_sub(1))
+        .map(|v| lower(v, boundary))
+}
+
+#[inline(always)]
+fn above(rel: impl Into<Line<usize>>, size: usize) -> Option<Span<usize>> {
+    raise(rel.into().end, ALIGN).map(|val| Span {
+        start: val,
+        count: size,
+    })
+}
```

Comment:
Do we still need these or can we use `memory::Address::{lower, raise}`

Reply:
Only, if we use checked integer arithmetic in `memory::Address`